vowpalwabbit/cb_explore_adf_bag.cc

```diff
@@ -0,0 +1,149 @@
+#include "cb_explore_adf_bag.h"
+#include "reductions.h"
+#include "cb_adf.h"
+#include "rand48.h"
+#include "bs.h"
+#include "gen_cs_example.h"
+#include "cb_explore.h"
+#include "explore.h"
+#include <vector>
+#include <algorithm>
+#include <cmath>
+
+// All exploration algorithms return a vector of id, probability tuples, sorted in order of scores. The probabilities
+// are the probability with which each action should be replaced to the top of the list.
+
+namespace VW {
+namespace cb_explore_adf {
+namespace bag {
+
+template <bool is_learn>
+void cb_explore_adf_bag::predict_or_learn_impl(LEARNER::multi_learner& base, multi_ex& examples)
+{
+  // Randomize over predictions from a base set of predictors
+  v_array<ACTION_SCORE::action_score>& preds = examples[0]->pred.a_s;
+  uint32_t num_actions = (uint32_t)examples.size();
+  if (num_actions == 0)
+  {
+    preds.clear();
+    return;
+  }
+
+  m_scores.clear();
+  for (uint32_t i = 0; i < num_actions; i++) m_scores.push_back(0.f);
+  m_top_actions.assign(num_actions, 0);
+  for (uint32_t i = 0; i < m_bag_size; i++)
+  {
+    // avoid updates to the random num generator
+    // for greedify, always update first policy once
+    uint32_t count = is_learn ? ((m_greedify && i == 0) ? 1 : BS::weight_gen(*m_random_state)) : 0;
+
+    if (is_learn && count > 0)
+      LEARNER::multiline_learn_or_predict<true>(base, examples, examples[0]->ft_offset, i);
+    else
+      LEARNER::multiline_learn_or_predict<false>(base, examples, examples[0]->ft_offset, i);
+
+    assert(preds.size() == num_actions);
+    for (auto e : preds) m_scores[e.action] += e.score;
+
+    if (!m_first_only)
+    {
+      size_t tied_actions = fill_tied(preds);
+      for (size_t i = 0; i < tied_actions; ++i) m_top_actions[preds[i].action] += 1.f / tied_actions;
+    }
+    else
+      m_top_actions[preds[0].action] += 1.f;
+    if (is_learn)
+      for (uint32_t j = 1; j < count; j++)
+        LEARNER::multiline_learn_or_predict<true>(base, examples, examples[0]->ft_offset, i);
+  }
+
+  m_action_probs.clear();
+  for (uint32_t i = 0; i < m_scores.size(); i++) m_action_probs.push_back({i, 0.});
+
+  // generate distribution over actions
+  exploration::generate_bag(
+      begin(m_top_actions), end(m_top_actions), begin_scores(m_action_probs), end_scores(m_action_probs));
+
+  exploration::enforce_minimum_probability(m_epsilon, true, begin_scores(m_action_probs), end_scores(m_action_probs));
+
+  sort_action_probs(m_action_probs, m_scores);
+
+  for (size_t i = 0; i < num_actions; i++) preds[i] = m_action_probs[i];
+}
+
+cb_explore_adf_bag::~cb_explore_adf_bag() {
+  m_action_probs.delete_v();
+}
+
+template <bool is_learn>
+void cb_explore_adf_bag::predict_or_learn(
+    cb_explore_adf_bag& data, LEARNER::multi_learner& base, multi_ex& examples)
+{
+  if (is_learn)
+    cb_explore_adf_base::learn(data, &cb_explore_adf_bag::predict_or_learn_impl<true>,
```

Comment:
Is there a way we can make the interface so that it is possible to call `data.learn(...)` instead of needing to call on the base class?

Reply:
We can, and I originally had it that way. But I was trying to figure out if I could bind some of the parameters in or use some other tricks so I could bypass the predict_or_learn function entirely and just pass base::predict/learn directly into the learner. So far I haven't found any way to do it, so I can change it back

Reply:
Also, it was a little weird that the line needs to be
data.predict(data, ...) (to get the type info)